openedx/core/djangoapps/content/course_structures/models.py

```diff
@@ -0,0 +1,73 @@
+import json
+
+from django.db import models
+from django.dispatch import receiver
+from celery.task import task
+from model_utils.models import TimeStampedModel
+
+from opaque_keys.edx.keys import CourseKey
+
+from xmodule.modulestore.django import modulestore, SignalHandler
+from xmodule_django.models import CourseKeyField
+
+class CourseStructure(TimeStampedModel):
+
+    course_id = CourseKeyField(max_length=255, db_index=True, verbose_name='Course ID')
+    version = models.CharField(max_length=255, blank=True, default="")
+
+    # Right now the only thing we do with the structure doc is store it and
+    # send it on request. If we need to store a more complex data model later,
+    # we can do so and build a migration. The only problem with a normalized
+    # data model for this is that it will likely involve hundreds of rows, and
+    # we'd have to be careful about caching.
+    structure_json = models.TextField(verbose_name='Structure JSON')
+
+    class Meta:
+        unique_together = ('course_id', 'version')
+
+
+def generate_course_structure(course_key):
+    """
+    Generates a course structure dictionary for the specified course.
+    """
+    course = modulestore().get_course(course_key, depth=None)
+    blocks_stack = [course]
+    blocks_dict = {}
+    while blocks_stack:
+        curr_block = blocks_stack.pop()
+        children = curr_block.get_children() if curr_block.has_children else []       
+        blocks_dict[unicode(curr_block.scope_ids.usage_id)] = {
+            "usage_key": unicode(curr_block.scope_ids.usage_id),
```

Comment:
Yeah, I'd lean towards just using storing `block_type` and `block_id`, rather than the full usage key, since the rest of the usage key is just the `course_key`. It might be good to also store the `definition_id`.